jplag.frontend.golang/src/main/java/de/jplag/golang/GoTokenConstants.java

```diff
@@ -0,0 +1,98 @@
+package de.jplag.golang;
+
+import de.jplag.TokenConstants;
+
+public interface GoTokenConstants extends TokenConstants {
+
+    // TOP LEVEL STRUCTURES
+
+    int PACKAGE = 2;
+    int IMPORT_CLAUSE = 3;
+    int IMPORT_CLAUSE_BEGIN = 4;
+    int IMPORT_CLAUSE_END = 5;
+    int IMPORT_DECLARATION = 6;
+    int ARRAY_BODY_BEGIN = 7;
+    int ARRAY_BODY_END = 8;
+    int STRUCT_DECLARATION = 9;
+    int STRUCT_BODY_BEGIN = 10;
+    int STRUCT_BODY_END = 11;
+    int INTERFACE_DECLARATION = 12;
+    int INTERFACE_BLOCK_BEGIN = 13;
+    int INTERFACE_BLOCK_END = 14;
+    int INTERFACE_METHOD = 15;
+    int TYPE_CONSTRAINT = 16;
+    int TYPE_ASSERTION = 17;
+    int MAP_BODY_BEGIN = 18;
+    int MAP_BODY_END = 19;
+    int SLICE_BODY_BEGIN = 20;
+    int SLICE_BODY_END = 21;
+    int NAMED_TYPE_BODY_BEGIN = 22;
+    int NAMED_TYPE_BODY_END = 23;
+    int MEMBER_DECLARATION = 24;
+
+    // FUNCTIONS AND METHODS
+
+    int FUNCTION_DECLARATION = 25;
+    int METHOD_DECLARATION = 26;
+    int RECEIVER = 27;
```

Comment:
This appears to me to open a relatively easy attack vector by simply refactoring a method to a function or vice versa. Why do you think it is worth having this separation?

Reply:
That's right, this transformation is probably very easy. I was too concerned about covering all kinds of sub-ASTs that I missed this. I need to keep that in mind.
The transformation of the invocations might still mess up the matching sequence, I suppose.

This is the relevant piece of the grammar. The receiver denotes the type on which the method shall be invocated, unlike other languages, where the method is associated with the surrounding type.
```
functionDecl: FUNC IDENTIFIER (signature block?);
methodDecl: FUNC receiver IDENTIFIER (signature block?);
```

Reply:
> The transformation of the invocations might still mess up the matching sequence, I suppose.
How would the tokenised invocations look like? I would assume independent of function or method that an invocation translates to an `INVOCATION` token.

If we want to prevent such attack, we had to unify the function and method declaration tokens as well as the receiver and function parameter tokens. However, I am not sure whether this is too much up-front optimisation to prevent one specific obfuscation attack.

Reply:
- Regardless of the function/method discussion, the arguments in the invocation may be any kind of expression, so if a argument expression produces a token, the order of the parameter can already easily be changed to avoid matches.
```golang
myFunc(getSomeObject(), []int{1})
//would produce INVOCATION ARG INVOCATION ARG SLICE() SLICE{ SLICE_ELEM }SLICE

myFunc([]int{1}, getSomeObject())
//would produce INVOCATION ARG SLICE() SLICE{ SLICE_ELEM }SLICE ARG INVOCATION
```

- I am not sure, either. At this point, generic type parameters are not supported by the grammar, so they are ignored. If the grammar and frontend are updated, the type arguments would sit right in the middle, so this optimization still would not work in that case, although that case may arguably be rare.
```golang
func (this myStruct) foo<T>(t T) {  }
//would produce FUNC PARAM TYPE_PARAM PARAM FUNC{ ...

func foo<T>(this myStruct, t T) {  }
//would produce FUNC TYPE_PARAM PARAM PARAM FUNC{ ...
```

Reply:
Following your argumentation that obfuscation in the parameters is possible anyways, I think we can close this discussion as it would probably lead to over engineering.